Tests/GRPCNIOTransportHTTP2Tests/HTTP2TransportTLSEnabledTests.swift

```diff
@@ -0,0 +1,402 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Crypto
+import Foundation
+import GRPCNIOTransportHTTP2Posix
+import NIOSSL
+import SwiftASN1
+import Testing
+import X509
+
+@Suite("HTTP/2 transport E2E tests with TLS enabled")
+struct HTTP2TransportTLSEnabledTests {
+  // - MARK: Test Utilities
+
+  // A combination of client and server transport kinds.
+  struct Transport: Sendable {
+    var server: ServerKind
+    var client: ClientKind
+
+    enum ClientKind: Sendable {
+      case posix(HTTP2ClientTransport.Posix.Config.TransportSecurity)
+    }
+
+    enum ServerKind: Sendable {
+      case posix(HTTP2ServerTransport.Posix.Config.TransportSecurity)
+    }
+  }
+
+  func executeUnaryRPCForEachTransportPair(
```

Comment:
The reason why I did it like this was because the important thing in these tests actually **is** the transport security config: incorrect config can make the TLS handshake fail, and it has to be all properly set up for the handshake to work. Each test has differences in how the config is set up. I wanted _that_ to be the focus of the tests - without it, the rest of the test is not really important: we're doing a unary request just to test the handshake but really it doesn't matter what we do after the handshake is completed.

To be clear, I'm not against refactoring all this - I'm sure it can be improved - but I'm not sure what the best alternative is. I could move the config to the parametrised argument, but it can get pretty large (and once we add support for NIOTS it will be worse):
```
@Test(arguments: [.posix(
  .tls(
    .defaults {
      $0.trustRoots = .certificates([.bytes(security.server.certificate, format: .der)])
      $0.serverHostname = "localhost"
    }
  )
)
],
[.posix(
  .tls(
    .defaults(
      certificateChain: [.bytes(security.server.certificate, format: .der)],
      privateKey: .bytes(security.server.key, format: .der)
    )
  )
)]
```

Reply:
I'm not suggesting we put all the config as a parameter for exactly the reason you pointed out, it's totally unwieldy! I'm suggesting we say the _type_ of transport we want to use and have the test call helpers which create the right transport config for you.

One of the things I don't like about these tests is that there's a lot of repetition in creating the config, and its not super obvious what the intent of the config is. There are only a few different skews of config:
- server, defaults
- server, mtls, with trust roots 
- server, mtls, with trust roots, and hostname
- client, defaults, and hostname
- client, mtls, with trust roots 
- client, mtls, with trust roots, and hostname

We can surely use this info to provide helpers that make these tests more readable and make the tls config more obvious?